File: server/antispam/antispam.go
package antispam

import (
	"strings"
	"sync"
	"time"
)

type submissionRecord struct {
	Timestamp time.Time
}

var (
	submissionCache sync.Map
)

func init() {
	// Periodic cleanup of expired cache entries (24h TTL)
	go func() {
		ticker := time.NewTicker(1 * time.Hour)
		for range ticker.C {
			cutoff := time.Now().Add(-24 * time.Hour)
			submissionCache.Range(func(key, value interface{}) bool {
				if rec, ok := value.(submissionRecord); ok {
					if rec.Timestamp.Before(cutoff) {
						submissionCache.Delete(key)
					}
				}
				return true
			})
		}
	}()
}

// SecurityResult encapsulates evaluation outcome
type SecurityResult struct {
	IsDuplicate       bool
	IsSuspicious      bool
	CompletionSeconds int
	Reasons           []string
}

// Evaluate evaluates submission integrity, honeypot, and timing heuristics
func Evaluate(responseID, startedAt, honeypot string) SecurityResult {
	var reasons []string
	isSuspicious := false

	// 1. Idempotency duplicate check
	_, isDuplicate := submissionCache.Load(responseID)

	// 2. Honeypot check
	if strings.TrimSpace(honeypot) != "" {
		isSuspicious = true
		reasons = append(reasons, "Honeypot field filled by automated bot")
	}

	// 3. Completion time heuristic
	completionSeconds := 60 // Default fallback
	if startedTime, err := time.Parse(time.RFC3339, startedAt); err == nil {
		diff := time.Since(startedTime).Seconds()
		if diff < 1 {
			diff = 1
		}
		completionSeconds = int(diff)
	}

	// 33 questions completed in under 40 seconds indicates bot or careless clicking
	if completionSeconds < 40 {
		isSuspicious = true
		reasons = append(reasons, "Completion time unrealistically fast (< 40s)")
	}

	return SecurityResult{
		IsDuplicate:       isDuplicate,
		IsSuspicious:      isSuspicious,
		CompletionSeconds: completionSeconds,
		Reasons:           reasons,
	}
}

// MarkComplete records responseID in the thread-safe cache
func MarkComplete(responseID string) {
	submissionCache.Store(responseID, submissionRecord{
		Timestamp: time.Now(),
	})
}

File: src/lib/security/antiSpam.ts
/**
 * Anti-Spam and Submission Integrity checks
 */

// In-memory idempotency cache (keyed by response_id) with 24-hour expiration
const completedSubmissions = new Map<string, { timestamp: number; success: boolean }>();

// Clean cache periodically to prevent memory leaks
if (typeof setInterval !== "undefined") {
  setInterval(() => {
    const cutoff = Date.now() - 24 * 60 * 60 * 1000;
    for (const [id, data] of completedSubmissions.entries()) {
      if (data.timestamp < cutoff) {
        completedSubmissions.delete(id);
      }
    }
  }, 60 * 60 * 1000);
}

export interface AntiSpamResult {
  isDuplicate: boolean;
  isSuspicious: boolean;
  reasons: string[];
  completionSeconds: number;
}

export function evaluateSubmissionSecurity(payload: {
  response_id: string;
  started_at: string;
  honeypot?: string;
}): AntiSpamResult {
  const reasons: string[] = [];
  let isSuspicious = false;

  // 1. Check idempotency (Duplicate check)
  const isDuplicate = completedSubmissions.has(payload.response_id);

  // 2. Honeypot check
  if (payload.honeypot && payload.honeypot.trim().length > 0) {
    isSuspicious = true;
    reasons.push("Honeypot field was filled by automated bot");
  }

  // 3. Completion time calculation and heuristic
  const startedTime = new Date(payload.started_at).getTime();
  const currentTime = Date.now();
  const completionSeconds = Math.max(
    1,
    Math.round((currentTime - (isNaN(startedTime) ? currentTime : startedTime)) / 1000)
  );

  // A 33-question survey completed in under 40 seconds indicates rapid bot or random clicking
  if (completionSeconds < 40) {
    isSuspicious = true;
    reasons.push(`Completion time unrealistically fast (${completionSeconds}s < 40s)`);
  }

  return {
    isDuplicate,
    isSuspicious,
    reasons,
    completionSeconds,
  };
}

export function markSubmissionComplete(response_id: string): void {
  completedSubmissions.set(response_id, {
    timestamp: Date.now(),
    success: true,
  });
}
